### PR TITLE
Add User entity and AuthenticationController:

### PR DESCRIPTION
**Pull request by DEz-1996**
Add User entity with DTOs and mapper
Add AuthenticationController with register endpoint Add FieldMatch annotation
Add RegistrationException and update GlobalExceptionHandler with it

### CODE REVIEW

File: src/main/java/online/book/store/validation/FieldMatchValidator.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+import online.book.store.dto.UserRegistrationRequestDto;
+
+public class FieldMatchValidator implements ConstraintValidator<FieldMatch, Object> {
+
+    @Override
+    public boolean isValid(Object object, ConstraintValidatorContext constraintValidatorContext) {
+        UserRegistrationRequestDto requestDto = (UserRegistrationRequestDto) object;
```
Comment (bhdnchui) [inline src/main/java/online/book/store/validation/FieldMatchValidator.java]: Let's specify FieldMatchValidator with UserRegistrationRequestDto type and avoid casting

File: src/main/java/online/book/store/exception/CustomGlobalExceptionHandler.java [hunk 2]
```diff
@@ -41,6 +41,12 @@ public ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundExcept
         return new ResponseEntity<>(body, HttpStatus.NOT_FOUND);
     }
 
+    @ExceptionHandler(RegistrationException.class)
+    public ResponseEntity<Object> handleRegistrationException(RegistrationException ex) {
+        Map<String, Object> body = Map.of("error", ex.getMessage());
+        return new ResponseEntity<>(body, HttpStatus.BAD_REQUEST);
```
Comment (bhdnchui) [inline src/main/java/online/book/store/exception/CustomGlobalExceptionHandler.java]: It's pretty same as a body in handleEntityNotFoundException method, let's avoid code duplication.

**Review by bhdnchui — CHANGES_REQUESTED** [review]
Good job! Just a few points to consider.

File: src/main/java/online/book/store/dto/UserRegistrationRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import org.hibernate.validator.constraints.Length;
+
+@Data
+@FieldMatch
+public class UserRegistrationRequestDto {
+    private static final int MIN_PASS_SIZE = 8;
+    private static final int MAX_PASS_SIZE = 20;
+
+    @NotBlank
+    @Email
```
Comment (bhdnchui) [inline src/main/java/online/book/store/dto/UserRegistrationRequestDto.java:16]: Do we need `@NotBlank` if we have `@Email`?

File: src/main/java/online/book/store/exception/CustomGlobalExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        body.put("status", HttpStatus.BAD_REQUEST);
+        List<String> errors = ex.getBindingResult().getAllErrors().stream()
+                .map(this::getErrorMessage)
+                .toList();
+        body.put("errors", errors);
+        return new ResponseEntity<>(body, headers, status);
+    }
+
+    @ExceptionHandler(EntityNotFoundException.class)
+    public ResponseEntity<Object> handleRegistrationException(EntityNotFoundException ex) {
```
Comment (bhdnchui) [inline src/main/java/online/book/store/exception/CustomGlobalExceptionHandler.java]: Please, rename. It handles `EntityNotFoundException`.

File: src/main/java/online/book/store/repository/UserRepository.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package online.book.store.repository;
+
+import java.util.Optional;
+import online.book.store.model.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+
```
Comment (bhdnchui) [inline src/main/java/online/book/store/repository/UserRepository.java:7]: Add annotation `@Repository`.

File: src/main/java/online/book/store/repository/BookRepository.java [hunk 6]
```diff
@@ -2,8 +2,6 @@
 
 import online.book.store.model.Book;
 import org.springframework.data.jpa.repository.JpaRepository;
-import org.springframework.stereotype.Repository;
 
-@Repository
```
Comment (bhdnchui) [inline src/main/java/online/book/store/repository/BookRepository.java]: Why did you remove this annotation?

File: src/main/java/online/book/store/validation/FieldMatchValidator.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+public class FieldMatchValidator implements
+        ConstraintValidator<FieldMatch,
+        UserRegistrationRequestDto> {
+
+    @Override
+    public boolean isValid(
+            UserRegistrationRequestDto requestDto,
+            ConstraintValidatorContext constraintValidatorContext
+    ) {
+        return requestDto.getPassword().equals(requestDto.getRepeatPassword());
```
Comment (bhdnchui) [inline src/main/java/online/book/store/validation/FieldMatchValidator.java]: Can getPassword() return null?
If it will return null, how are we going to handle NPE?
Just use [Objects#equals](https://www.baeldung.com/java-comparing-objects#equals-static).

**Review by bhdnchui — CHANGES_REQUESTED** [review]
Good job!
Just a few comments to process
Also, pay attention that your PR contains a lot of changes from previous tasks, please fix it.

File: src/main/java/online/book/store/service/imp/UserServiceImp.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+import online.book.store.exception.exceptions.RegistrationException;
+import online.book.store.mapper.UserMapper;
+import online.book.store.model.User;
+import online.book.store.repository.UserRepository;
+import online.book.store.service.UserService;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class UserServiceImp implements UserService {
```
Comment (d1sam) [inline src/main/java/online/book/store/service/imp/UserServiceImp.java]:
```suggestion
public class UserServiceImpl implements UserService {
```

File: src/main/java/online/book/store/validation/FieldMatch.java [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+import jakarta.validation.Payload;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Constraint(validatedBy = FieldMatchValidator.class)
+@Target({ElementType.TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface FieldMatch {
```
Comment (d1sam) [inline src/main/java/online/book/store/validation/FieldMatch.java]:
```suggestion
public @interface PasswordsFieldsMatch {
```

Comment (d1sam) [reply src/main/java/online/book/store/validation/FieldMatch.java]: because you check only passwords in your `FieldMatchValidator`

File: src/main/java/online/book/store/controller/AuthenticationController.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+@RequestMapping("/auth")
+@RequiredArgsConstructor
+@Tag(name = "User authentication", description = "Endpoints for authentication users")
+public class AuthenticationController {
+    private final UserService userService;
+
+    @PostMapping("/registration")
+    @Operation(summary = "Register new user", description = "Register new user")
+    public UserResponseDto register(@RequestBody @Valid UserRegistrationRequestDto request)
+            throws RegistrationException {
```
Comment (d1sam) [inline src/main/java/online/book/store/controller/AuthenticationController.java]:
```suggestion {
```

File: src/main/java/online/book/store/dto/CreateBookRequestDto.java [hunk 11]
```diff
@@ -1,15 +1,33 @@
...
+    private static final int MAX_STRING_LENGTH = 255;
+    private static final String STRING_SIZE_MSG = "length must be between 5 and 255 chars";
+    private static final String ISBN_SIZE_MSG = "length must be 10 or 13 chars";
+
+    @NotNull
+    @Size(min = MIN_STRING_LENGTH, max = MAX_STRING_LENGTH, message = STRING_SIZE_MSG)
     private String author;
+    @NotNull
+    @Size(min = MIN_STRING_LENGTH, max = MAX_STRING_LENGTH, message = STRING_SIZE_MSG)
+    private String title;
```
Comment (d1sam) [inline src/main/java/online/book/store/dto/CreateBookRequestDto.java:23]: for author and title can be passed blank value "           " with only spaces

**Review by d1sam — CHANGES_REQUESTED** [review]
Good job, but some changes should be done.

**Review by d1sam — APPROVED** [review]
Well done!
